### PR TITLE
feat: refactor dockfile to support environment variables

### PR DESCRIPTION
**Pull request by jasonyic**
## Feature request
existing `Dockerfile` is outdated and does not support environment variable to overwrite default configuration value or command line options. 

## Changes
refactor `Dockerfile` to include `entrypoint` script to provide the flexibility to overwrite default `harmony.conf` with environment variable. only the default configuration supported by the `harmony` release can be overwritten by the env, and new configuration WON"T be added if supported in different harmony version. e.g  `KMSEnabled` in release `2.5.3` will not be added into the configuration in release `2.5.2`, when env `HARMONY_BLSKEYS_KMSENABLED` is used.

`Env` variable needs to be prefixed with `HARMONY_` and followed by `Group Name` and `Config Name`. e.g  to overwrite `NetworkType` under `Network` section in `harmony.conf`, the following environment variable should be used to overwrite default value `testnet`:
 ```
HARMONY_NETWORK_NETWORKTYPE=mainnet
```
`harmony.conf` initialised with default values for `testnet` and all configuration can be overwritten in environment variables.  this will provide much more flexibility when docker is used to host the harmony client, e.g. Kubernetes.

a `docker-compose.yaml` is also provided for local testing

update `makefile` to include `build-docker` target

feature request: #4211

### CODE REVIEW

**Review by LeoHChen — COMMENTED** [review]
do. you have any test log of the new docker image?

File: scripts/docker/docker-entrypoint.sh [hunk 1]
```diff
@@ -0,0 +1,93 @@
+#!/bin/bash
+set -e
```
Comment (diego1q2w) [inline scripts/docker/docker-entrypoint.sh:2]: Could you add a comment at the top explaining what this script does? I think is important to enphasise that it only replaces existing values but won't add new ones.

File: scripts/docker/docker-entrypoint.sh [hunk 2]
```diff
@@ -0,0 +1,93 @@
...
+fi
+
+echo "#############################"
+echo "### Harmony Configuration ###"
+echo
+cat ${CONFIG_FILE}
+echo
+echo "#############################"
+
+exec yes no | "$@"
```
Comment (diego1q2w) [inline scripts/docker/docker-entrypoint.sh:93]: what does this do?

**Review by diego1q2w — COMMENTED** [review]
Left a couple of comments but otherwise looks good

File: scripts/docker/docker-entrypoint.sh (hunk 1 above)
Comment (jasonyic) [reply scripts/docker/docker-entrypoint.sh:2]: updated. please review

File: scripts/docker/docker-entrypoint.sh (hunk 2 above)
Comment (jasonyic) [reply scripts/docker/docker-entrypoint.sh]: was used to auto-answer the version update prompt. not applicably and removed.

**Comment by jasonyic** [comment]
> do. you have any test log of the new docker image?

I did try to update the `harmony-test` to cater for the rpc testing for docker container, but quite some effort to rewrite the test cases  when using docker container with testnet. we did run the load testing using k6 to test the docker container build on this PR, and here is the test result on 8CPUs/24Gi RAM

we did experience issue on `hmyv2_call` which can not go over 200 RPS.

APIs in test
```
  {"jsonrpc":"2.0","id":1,"method":"hmyv2_inSync","params":[]},
  {"jsonrpc":"2.0","id":1,"method":"hmyv2_getPendingCXReceipts","params":[]},
  {"jsonrpc":"2.0","id":1,"method":"hmyv2_getPoolStats","params":[]},
  {"jsonrpc":"2.0","id":1,"method":"hmyv2_blockNumber","params":[]},
  {"jsonrpc":"2.0","id":1,"method":"hmyv2_getEpoch","params":[]},
  {"jsonrpc":"2.0","id":1,"method":"hmyv2_getBlockTransactionCountByNumber","params":[1]},
  {"jsonrpc":"2.0","id":1,"method":"hmyv2_getBalance","params":["0x5D64c230ed2e39E0157E39Aed2E4D5D5DEc1C3f0"]},
  {"jsonrpc":"2.0","id":1,"method":"hmyv2_getBlocks","params":[1,2,{"withSigners": false, "fullTx": false,"inclStaking": false}]},
  {"jsonrpc":"2.0","id":1,"method":"hmyv2_call","params":[{"to": "0x5575821Ebf7427ed81907a373cF7de7Dd4403C99"},0]},
  {"jsonrpc":"2.0","id":1,"method":"hmyv2_getTransactionReceipt","params":["0xd324cc57280411dfac5a7ec2987d0b83e25e27a3d5bb5d3531262387331d692b"]}
```
K6 Result:
```
  scenarios: (100.00%) 1 scenario, 5000 max VUs, 10m30s max duration (incl. graceful stop):
           * contacts: 1500.00 iterations/s for 10m0s (maxVUs: 2000-5000, gracefulStop: 30s)


running (10m00.2s), 0000/2128 VUs, 899691 complete and 0 interrupted iterations
contacts ✓ [======================================] 0000/2128 VUs  10m0s  1500 iters/s

     █ Infura - Harmony - READ/GET

       ✓ is status 200
       ✓ verify rpc resp
       ✓ verify rpc resp - no err

     checks.........................: 100.00% ✓ 2699073     ✗ 0     
     data_received..................: 827 MB  1.4 MB/s
     data_sent......................: 212 MB  354 kB/s
     dropped_iterations.............: 310     0.51648/s
     group_duration.................: avg=221.04ms min=211.12ms med=214.5ms  max=3.29s    p(90)=219.34ms p(95)=223.56ms
     http_req_blocked...............: avg=6.8µs    min=800ns    med=2.13µs   max=86.59ms  p(90)=3.69µs   p(95)=4.42µs  
     http_req_connecting............: avg=3.35µs   min=0s       med=0s       max=86.55ms  p(90)=0s       p(95)=0s      
     http_req_duration..............: avg=220.9ms  min=211.04ms med=214.38ms max=3.29s    p(90)=219.18ms p(95)=223.38ms
       { expected_response:true }...: avg=220.9ms  min=211.04ms med=214.38ms max=3.29s    p(90)=219.18ms p(95)=223.38ms
     http_req_failed................: 0.00%   ✓ 0           ✗ 899691
     http_req_receiving.............: avg=38.45µs  min=6.68µs   med=21.39µs  max=151.72ms p(90)=39.93µs  p(95)=49.4µs  
     http_req_sending...............: avg=29.25µs  min=3.81µs   med=10.45µs  max=219.15ms p(90)=21.55µs  p(95)=30.98µs 
     http_req_tls_handshaking.......: avg=0s       min=0s       med=0s       max=0s       p(90)=0s       p(95)=0s      
     http_req_waiting...............: avg=220.83ms min=211.01ms med=214.33ms max=3.29s    p(90)=219.12ms p(95)=223.29ms
     http_reqs......................: 899691  1498.944126/s
     iteration_duration.............: avg=221.06ms min=211.13ms med=214.52ms max=3.29s    p(90)=219.36ms p(95)=223.59ms
     iterations.....................: 899691  1498.944126/s
     vus............................: 2128    min=2000      max=2128
     vus_max........................: 2128    min=2000      max=2128
```

File: scripts/docker/docker-entrypoint.sh (hunk 1 above)
Comment (diego1q2w) [reply scripts/docker/docker-entrypoint.sh:2]: Didn't see the comment, where did you add it?

Comment (jasonyic) [reply scripts/docker/docker-entrypoint.sh:2]: added in the PR description under `changes`

**Review by diego1q2w — APPROVED** [review]
Looks good to me unless @LeoHChen or @sophoah  would like to make any further comments it's good to go

File: scripts/docker/docker-entrypoint.sh (hunk 1 above)
Comment (jasonyic) [reply scripts/docker/docker-entrypoint.sh:2]: thank you

**Comment by jasonyic** [comment]
After gave it a second thought. I think the configuration file should be used as the single source of truth when configuring the node. `env` does seem dynamic, but will cause quite some extra work to maintain. From operational perspective, debugging & trouble shooting will be hard when using env. I submitted another PR https://github.com/harmony-one/harmony/pull/4239 to rely on configuration file to boot the node. detailed use cases also added in README.md.

**Pull request closed without merging**